service/matching/taskWriter.go

```diff
@@ -170,17 +173,12 @@ func (w *taskWriter) allocTaskIDs(count int) ([]int64, error) {
 func (w *taskWriter) appendTasks(
 	tasks []*persistencespb.AllocatedTaskInfo,
 ) (*persistence.CreateTasksResponse, error) {
-
 	resp, err := w.tlMgr.db.CreateTasks(tasks)
-	switch err.(type) {
-	case nil:
-		return resp, nil
-
-	case *persistence.ConditionFailedError:
-		w.tlMgr.Stop()
+	if err != nil && errIndicatesForeignLessee(err) {
+		w.tlMgr.signalFatalProblem(w.tlMgr.taskQueueID)
 		return nil, err
-
-	default:
+	}
```

Comment:
comparing 1st option vs 2nd option, i would say 1st one is much easier to read?
```go
switch err.(type) {
case nil:
	return resp, nil
case *persistence.ConditionFailedError:
	w.tlMgr.signalFatalProblem(w.tlMgr.taskQueueID)
	return nil, err
default:
	return nil, err
}
```

vs

```go
func errIndicatesForeignLessee(err error) bool {
	var condfail *persistence.ConditionFailedError
	return errors.As(err, &condfail)
}

...
if err != nil && errIndicatesForeignLessee(err) {
	w.tlMgr.signalFatalProblem(w.tlMgr.taskQueueID)
	return nil, err
}
if err != nil {
	return nil, err
}
return resp, nil
```

Reply:
They're not functionally equivalent but the second is more clear for me anyway, that's why I changed it.

- errIndicatesForeignLessee says what actually happened rather than just comparing a seemingly random error type
- It also works with any error that has ConditionFailedError in the causal chain so if the persistence layer starts wrapping things this code continues to work

Reply:
>seemingly random error type

actually, the error type `persistence.ConditionFailedError` means conditional write to DB fails, e.g. ownership lost
there are contracts of what error types mean

>It also works with any error that has ConditionFailedError in the causal chain so if the persistence layer starts wrapping things this code continues to work

there is no causal chain, all errors from dependency must be flat, if there is causal chain, that is essentially making business logic to handle errors from different e.g. persistence impl, e.g. SQL can return some error while CQL can return another CQL specific error. 
dependency layer MUST do error conversion to error type which business logic understand

Reply:
> there are contracts of what error types mean
Where? None of `taskQueueDB.RenewLease`, `persistence.TaskManager.LeaseTaskQueue`, `persistence.taskManagerImpl.LeaseTaskQueue`, or the `persistence.ConditionFailedError` mention anything about the type of error returned. Obviously I can read the code. but the reason I'm careful here is because I can't find any existing promise or contract.

> there is no causal chain
Today. The point is to be resilient to change here and at the same time allow for evolution (through error.Is and error.As) by the lower layer(s). It's sort of similar to coupling to an interface rather than a concrete type.

I understand what you're saying about encapsulating layers. But let's try to avoid painting ourselves into a corner where there can only be a single error type (or worse, error _instance_) in a layer.

Reply:
I don't think I truly understand your objection here. The code _works_ as is, I don't think that's at issue ... so can you say which aspect of it you find problematic?

Reply:
>Where? None of taskQueueDB.RenewLease, persistence.TaskManager.LeaseTaskQueue, persistence.taskManagerImpl.LeaseTaskQueue, or the persistence.ConditionFailedError mention anything about the type of error returned. Obviously I can read the code. but the reason I'm careful here is because I can't find any existing promise or contract.

doc should be improved, but if you take a look at the impl, `ConditionFailedError` should mean CAS fail, and for that API, that means ownership lost


>Today. The point is to be resilient to change here and at the same time allow for evolution (through error.Is and error.As) by the lower layer(s). It's sort of similar to coupling to an interface rather than a concrete type.

>I understand what you're saying about encapsulating layers. But let's try to avoid painting ourselves into a corner where there can only be a single error type (or worse, error instance) in a layer.

>I don't think I truly understand your objection here. The code works as is, I don't think that's at issue ... so can you say which aspect of it you find problematic?

in terms of functionality, nothing breaks; what i am objecting here is usage of error.Is and error.As.
the usage will inevitably force business logic layer to deal with different dependencies' error (error type).

when the dependency return an error, the error type must be part of the contract.

one more thing, 
```go
var condfail *persistence.ConditionFailedError
if errors.As(err, &condfail) {
    ...
}
var otherErr *some.OtherErrorType
if errors.As(err, &otherErr) {
    ...
} 
if err != nil {
    ...
}
```
is verbose, e.g. especially we need to handle multiple error types.

switch is much more cleaner
```go
switch err.(type) {
case nil:
	...
case *persistence.ConditionFailedError:
	...
default:
	...
}
```

Reply:
> inevitably force business logic layer to deal with different dependencies' error (error type).

Can you walk me through that? I don't think I buy the argument. It's certainly not present in this PR.

I agree that the verbosity isn't great but this has been the recommended way to test errors for a couple of years now. And the problem you describe with repeated type tests - that's also not present in this PR.

Reply:
>but this has been the recommended way to test errors for a couple of years now

depends on the situation, for lib / SDK, using errors.As or errors.Is is probably reasonable, but server is a different story.

>And the problem you describe with repeated type tests - that's also not present in this PR.

yes and no, there is not too much of verbosity issue, but you are create yet another way to handling errors which i do not believe

Reply:
> but server is a different story

how? 

> ... handling errors which i do not believe

Let's talk about this directly. It's a big wast of time to go through these arguments and in the end you just don't believe in something.

Reply:
>how?

server's dependency wrapper must convert the dependency specific error to error types business logic understand.

>Let's talk about this directly. It's a big wast of time to go through these arguments and in the end you just don't believe in something.

let us talk, i do not understand why errors.As / errors.Is is required

Reply:
>how?

server itself (dependency wrapper) must proactively convert dependency error to error type(s) business logic can handle (business logic / persistence contract)

image presenting a random error to business logic layer, what should business logic do? bubble up the error to API caller? 

in practice, errors are categorized / converted by each dependency wrapper / driver, e.g.:
* NotFound: target is gone / does not exists, converted from e.g. `sql.ErrNoRows`, `gocql.ErrNotFound`
* CAS related error e.g. `WorkflowConditionFailedError`, `ShardOwnershipLostError`, converted by parsing returned DB record
* Internal server error: server error, e.g. cassandra returning some random error that persistence layer does not understand.

above contract means all errors returned by dependency layer are of known type, so there is no real need to use errors.Is and errors.As, since <- is relatively verbose and we do not need to see the underlying error within (because there is no error within)